src/emmy/numerical/ode.cljc

```diff
@@ -2,96 +2,171 @@
+  (:require #?@(:cljs
+                [["odex" :as o]
+                 [emmy.util :as u]])
+            #?(:clj
+               [clojure.core.async :as a])
             [emmy.expression.compile :as c]
             [emmy.structure :as struct]
             [emmy.util.stopwatch :as us]
+            [emmy.value :as v]
             [taoensso.timbre :as log])
   #?(:clj
      (:import (org.apache.commons.math3.ode.nonstiff GraggBulirschStoerIntegrator)
               (org.apache.commons.math3.ode FirstOrderDifferentialEquations)
+              (org.apache.commons.math3.ode ExpandableStatefulODE)
               (org.apache.commons.math3.ode.sampling StepHandler))))
 
-#?(:cljs
-   (def ^:private near? (v/within 1e-8)))
 
-#?(:clj
-   (defn- round-up
-     "Returns `n` rounded up to the nearest multiple of `step-size`. the returned
-  value will always equal `0`, mod `step-size`"
-     [n step-size]
-     (let [offset (mod n step-size)]
-       (if (pos? offset)
-         (-> (- n offset)
-             (+ step-size))
-         n))))
 
-#?(:clj
-   (defn step-handler
-     "Generates a StepHandler instance that can be attached to an integrator.
+(def ^:private near? (v/within 1e-8))
+(def ^:private default-epsilon 1e-8)
+
+(defn stream-integrator
+  "Produces a function, monotonic in its single numeric argument, 
+   that represents the integral of the function f' given the initial 
+   data $y_0 = f(x_0)$ and an options dictionary (presently containing
+   the tolerance for error $\\epsilon$, but eventually also selecting
+   from a menu of integration techniques). 
+   
+   This is done by creating an adaptive step-size ODE solver, and 
+   advancing its steps as needed to supply function values. (This 
+   architecture accounts for why the arguments to f must be presented
+   in order). Old solution segments are discarded. The goal of this 
+   approach is to avoid the requirement of supplying an upper limit
+   to the integration. At the cost of requiring monotonic arguments
+   to f, the integrated function can essentially be used forever
+   without accumulating unbounded state.
+   
+   The returned function may be called with no arguments to shut down
+   the integration, allowing for the final reclamation of its resources.
+
+   When the ODE solver is provided by Java, it may be necessary to 
+   use an auxiliary thread to enable this style of flow control.  If
+   JavaScript, we expect the solver to provide a generator of solution
+   segments."
+  [f' x0 y0 {:keys [epsilon] :or {epsilon default-epsilon}}]
+  (let [dimension (count y0)]
+    #?(:clj
+       (let [gbs               (GraggBulirschStoerIntegrator. 0. 1. (double epsilon) (double epsilon))
+             ode               (ExpandableStatefulODE.
+                                (reify FirstOrderDifferentialEquations
+                                  (computeDerivatives
+                                    [_ x y out]
+                                    (let [y' (f' x y)]
```

Comment:
adding a note for us to discuss — I was thinking while writing the animations that we might make the ODE interface require a precompiled function 

```clj
(f' x y out)
```

where `f'` is responsible for mutating the array. Then on the way in, we can wrap an `f'` that returns a clojure sequence:

```clj
(defn ->arr
  [^doubles arr ^Iterable xs]
  (let [it (.iterator xs)]
    (loop [i 0]
      (when (.hasNext it)
        (aset arr i (double (.next it)))
        (recur (unchecked-inc i))))))

(defn ->raw [f]
  (fn [x y out]
    (->arr out (f x y))))
```

I benchmarked a `->raw*` that used the `doseq` (which is how I initially wrote it too, before polling Clojurians Slack and getting the above solution), and found that `->arr` above is ~10-15x faster in my microbenchmark:

```clj
(defn ->arr
  [^doubles arr ^Iterable xs]
  (let [it (.iterator xs)]
    (loop [i 0]
      (when (.hasNext it)
        (aset arr i (double (.next it)))
        (recur (unchecked-inc i))))))

(defn ->raw [f]
  (fn [x y out]
    (->arr out (f x y))))

(defn ->raw* [f]
  (fn [x y ^doubles out]
    (let [y' (f x y)]
      (doseq [i (range 0 (alength out))]
        (aset out i (double (nth y' i)))))))

(def out (double-array 3))

(let [f1 (->raw  (fn [_ _] [1 2 3]))
      f2 (->raw* (fn [_ _] [1 2 3]))
      n  100000000]
  (time
   (dotimes [_ n]
     (f1 1 2 out)))
  (time
   (dotimes [_ n]
     (f2 1 2 out))))
;; Elapsed time: 1016.89225 msecs
;; Elapsed time: 16735.008792 msecs
```

Reply:
The only other change I would suggest is that we allow the user to specify stateful parameters that can change. In scmutils, these come in via the first layer of the "state derivative"; I found that it worked really nicely to

- take a vector of parameters as an extra argument, like `(f' x y params out)`, if the function needs them
- Take an ATOM to supply the params, and internally call `(f' x y (.-state state-atom) out)`.

The reason to call `(.-state state-atom)` vs `@` aka `deref` is that accessing `-state` directly on the JS side will get around Reagent's "do I re-render" watcher, which checks for deref calls.

Reply:
Definitely worth thinking about. At the bottom, the solvers now both work with primitive arrays, and we could save an allocation by allowing the user of the primitive system to mutate an out vector instead of returning something. That's kind of a defect of the way I set up odex-js, though, since they are expected to return an array

Reply:
@littleredcomputer that's true re: Odex!! but to play nice with Odex, we can allocate a single `out` array for return values, pass it into `f'` and then continually return the same `out` array to Odex.

Reply:
I am convinced, so I will rework things to allow a preallocated array.

Reply:
Rereading what you have written carefully, let me see if I have it right.

- `stream-integrator` will insist on being called with a "raw function", i.e., one with the signature [x y out] where y and out are [D. 
- Users of the classic SICM interface will have their derivatives wrapped into raw form.
- Since the compiler is hidden inside the SICM-style interface, you will have the privilege of relying on this so the compiler can produce a derivative that fulfills the raw contract, skipping the wrapping when the compiler is used.
- I will change the expected function signature in ODEX to raw, so that we will lose no performance in the JS case. I'm making a 3.0.0 version, so I'm permitted this breaking change. 
- I could then add an option requesting the old calling convention 
- Or I could be "magic" and say if f' returns "something" then that's the value else the contents of `out`, so it magically works with old functions and new. But I have talked myself out of this

Reply:
On the subject of `->arr`: Currently, what goes into the fixed array after the state derivative is called is the result of flattening the state increment. `flatten` produces a LazySeq, which, unfortunately, is not considered `Iterable` in cljs, so that code snippet doesn't work.

Further study revealed that the long pole time wise is actually the `flatten` itself, about 10x, which dwarfs the difference between `doseq` and `->arr`. 

Casting about for a solution that was better than the naive approach while being portable to cljs without hacking, I decided to try `clojure.core.reducers` and the good news is that the flatten in the reducers library is much more performant in the microbenchmark than `clojure.core/flatten`. This coupled with an `aset` in the reducing step seems like a fair compromise

```clojure
(require '[clojure.core.reducers :as r])

(defn flatten-into-primitive-array [a s]
  (let [ix (atom -1)
        asetter aset-double]
    (r/reduce (fn [a t]
                (asetter a (swap! ix unchecked-inc) t)
                a)
              a
              (r/flatten s))))

(let [a (double-array 7)
      s [1. [2. 3. 4.] [5. 6. 7]]
      n 10000000]
  (time
   (dotimes [_ n]
     (->arr a (flatten s))))
  (time
   (dotimes [_ n]
     (flatten-into-primitive-array a s))))

"Elapsed time: 21511.873042 msecs"
"Elapsed time: 6022.566583 msecs"
```

so while your solution was down to ~600msec, it was exclusive of flatten. For this variant of the microbenchmark, we are flattening something which looks very much like a Sussmanian Derivative

Reply:
this has been implemented and further simplifies ode.clj, no longer requiring different versions of `state->array` for clj and class

Reply:
Returning now to the subject of atoms, wherein you request another parameter in the derivative to hold the time-varying values, would it not be possible to just close over the atoms you need without involving the ODE interface?

Reply:
yes, it would for sure, if you tweak `compile-state-fn` to expect an atom and then internally deref it... we can look at these styles when we get the animation example up and start hacking on that.